ddtrace/contrib/internal/ray/span_manager.py

```diff
@@ -0,0 +1,263 @@
+import atexit
+from contextlib import contextmanager
+from itertools import chain
+import threading
+import time
+from typing import Dict
+from typing import Optional
+from typing import Tuple
+from typing import Union
+
+from ray.dashboard.modules.job.common import JobInfo
+
+from ddtrace import config
+from ddtrace import tracer
+from ddtrace._trace.context import Context
+from ddtrace._trace.span import Span
+from ddtrace.constants import ERROR_MSG
+from ddtrace.constants import SPAN_KIND
+from ddtrace.ext import SpanKind
+
+from .constants import DD_PARTIAL_VERSION
+from .constants import DD_WAS_LONG_RUNNING
+from .constants import RAY_COMPONENT
+from .constants import RAY_JOB_MESSAGE
+from .constants import RAY_JOB_STATUS
+from .constants import RAY_STATUS_FAILED
+from .constants import RAY_STATUS_FINISHED
+from .constants import RAY_STATUS_RUNNING
+from .constants import RAY_SUBMISSION_ID_TAG
+from .utils import _inject_ray_span_tags_and_metrics
+
+
+@contextmanager
+def long_running_ray_span(
+    span_name: str,
+    service: str,
+    span_type: str,
+    resource: Optional[str] = None,
+    child_of: Optional[Union[Span, Context]] = None,
+    activate: bool = True,
+):
+    """Context manager that handles Ray span creation and long-running span lifecycle"""
+    with tracer.start_span(
+        name=span_name, service=service, resource=resource, span_type=span_type, child_of=child_of, activate=activate
+    ) as span:
+        span.set_tag_str(SPAN_KIND, SpanKind.CONSUMER)
+        _inject_ray_span_tags_and_metrics(span)
+        start_long_running_span(span)
+
+        try:
+            yield span
+        except BaseException as e:
+            raise e
+        finally:
+            stop_long_running_span(span)
+
+
+class RaySpanManager:
+    def __init__(self) -> None:
+        self._timers: Dict[str, threading.Timer] = {}
+        # {submission_id: {(trace_id, span_id): Span}}
+        self._job_spans: Dict[str, Dict[Tuple[int, int], Span]] = {}
+        # {submission_id: (Span)}
+        self._root_spans: Dict[str, Span] = {}
+        self._lock: threading.Lock = threading.Lock()
```

Comment:
I missed this before, but @mabdinur called out, about us using a forksafe lock here. Did that conversation ever get resolved?

Definitely safer to have this as a `ddtrace.internal.forksafe.Lock`

Reply:
I did not resolve it because I was waiting for your answer !